src/tree/decisiontree.ts

```diff
@@ -0,0 +1,296 @@
+            this.nodes[current_node_id].n_samples
+        )
+      }
+      class_probabilities.push(current_class_probabilities)
+    }
+    return class_probabilities
+  }
+  predict(samples: number[][]): int[] {
+    if (!this.is_built) {
+      throw new Error(
+        'Decision tree must be built with BuildTree method before predictions can be made.'
+      )
+    }
+    let leaf_node_ids = this.GetLeafNodes(samples)
+    let class_predictions = []
+
+    for (let n_sample = 0; n_sample < leaf_node_ids.length; n_sample++) {
+      let current_node_id = leaf_node_ids[n_sample]
+      class_predictions.push(argMax(this.nodes[current_node_id].value))
+      this.nodes[current_node_id].value
+    }
+    return class_predictions
+  }
+}
+
+export class DecisionTreeClassifier extends ClassifierMixin {
+  splitter_?: Splitter
+  stack_: NodeRecord[] = []
+  min_samples_leaf: int
+  max_depth_: int
+  min_samples_split: int
+  min_impurity_split_: number
+  n_features_: int = 0
+  tree_: DecisionTree
+  criterion_: ImpurityMeasure
+  max_features_: int
+  max_features_method: 'log2_method' | 'sqrt_method' | 'all_method'
+  feature_data_: number[][] = []
+  label_data_: number[] = []
+
+  constructor({
+    criterion = 'gini',
+    max_depth = Number.POSITIVE_INFINITY,
+    min_samples_split = 2,
+    min_samples_leaf = 1,
+    max_features = -1,
+    max_features_method = 'all_method',
+    min_impurity_split = 0.0
+  } = {}) {
+    super()
+    this.criterion_ = criterion as any
+    this.max_depth_ = max_depth
+    this.min_samples_split = min_samples_split
+    this.min_samples_leaf = min_samples_leaf
+    this.max_features_ = max_features
+    this.max_features_method = max_features_method as any
+    this.min_impurity_split_ = min_impurity_split
+    this.tree_ = new DecisionTree()
+  }
+
+  predict(feature_data: number[][]) {
+    return this.tree_.predict(feature_data)
+  }
+
+  predictProba(feature_data: number[][]) {
+    return this.tree_.predictProba(feature_data)
+  }
+  fit(feature_data: number[][], label_data: int[], samples_subset?: number[]) {
+    this.feature_data_ = feature_data
+    this.label_data_ = label_data
+
+    let new_samples_subset = samples_subset || []
+
+    // CheckNegativeLabels(label_data_ptr);
+    this.max_features_ = SetMaxFeatures(
+      this.max_features_,
+      this.max_features_method,
+      feature_data
+    )
+
+    this.splitter_ = new Splitter(
+      feature_data,
+      label_data,
+      this.min_samples_leaf,
+      this.criterion_,
+      this.max_features_,
+      new_samples_subset
+    )
+
+    // put root node on stack
+    let root_node: NodeRecord = {
+      start: 0,
+      end: this.splitter_.sample_map_.length,
+      depth: 0,
+      impurity: 0,
+      n_samples: this.splitter_.sample_map_.length,
+      parent_id: -1,
+      is_left: false,
+      value: []
+    }
+    this.stack_.push(root_node)
+
+    let is_root_node = true
+
+    while (this.stack_.length !== 0) {
```

Comment:
I agree. That was actually the approach that I started with. 

I'm basing a lot of this code off of the sklearn code [here](https://github.com/scikit-learn/scikit-learn/blob/7e1e6d09bcc2eaeba98f7e737aac2ac782f0e5f1/sklearn/tree/_tree.pyx) , which uses this approach because they have a `DepthFirstBuilder` and a `BestFirstBuilder`. So basically you can recurse down the tree by which node has the worst/best impurity, or by depth, and so they do this PriorityQueue approach instead of the easier recursion.